### ISSUE TITLE
Do not require  deps.edn when using local bbin install 

### ISSUE THREAD

**Issue by publicimageltd** [babashka/bbin#78]
I have been working on a babashka project with the following `bb.edn` (the script's name is `secrets`):

```clojure
{:paths ["src"] ;; library
 :deps {com.taoensso/timbre {:mvn/version "6.4.0"}}
 :bbin/bin {secrets {:main-opts ["-m" "secrets.core/-main"]}}
}
```

Installation via `bbin install .` works, but when I call the resulting script, a classpath not found error is raised. So I added an empty `deps.edn`:

```clojure
{:paths ["src"]}
```

Now the reinstalled script works. But why is that necessary? The `deps.edn` is useless, since I am already declaring the dependencies in the `bb.edn` file and do not call `clj`. So I expected `bbin install` to work autonomously. 

So if I did not miss something, I think it would be better to drop this dependency. Not least because jacking into the script with CIDER causes unnecessary confusion (I have to choose "clojure or babashka?"). If that will not be changed, I would recommend at least to document that a `deps.edn` is also necessary.

**Comment by borkdude** [comment]
cc @rads

**Comment by NoahTheDuke** [comment]
I've run into this issue another way: I have a deps.edn that has a dependency that doesn't work with babashka, so in my bb.edn, I don't depend on the local deps.edn, I just specify `:paths` and use `deps/add-lib` in my script to bring in a babashka compatible version:

```clojure
#?(:bb (do (require '[babashka.deps :as deps])
           (deps/add-deps '{:deps {io.github.babashka/tools.bbuild
                                   {:git/sha "f5a4acaf25ec2bc5582853758ba81383fff5e86b"}}}))
```

If I run the script with babashka, this works, but if I run it with `bbin`'s generated script, it doesn't because it pulls in the incompatible version through the deps.edn first.

**Comment by teodorlu** [comment]
I had a look at this, made a reproduction, and wrote a very long summary of what I learned:

https://github.com/teodorlu/bbin-testdata/tree/72934325b97ce431c57ef0c581b4ed1aeedf70af/fail1/README.md

Summary of what I learned:

1. The present version of bbin uses the `--deps-root` argument when calling `bb`, which crashes if `deps.edn` is not present in the script root folder
2. From my limited testing, `--deps-root` does not appear to be necessary, setting `--config` to the script's `deps.edn` file or `bb.edn` file may be sufficient.
3. A modified bbin that uses `bb --config SCRIPT_DEPS_FILE` will crash if the `deps.edn` file or `bb.edn` file does not set `:paths` explicitly, `:paths ["src"]` is not inferred.

For 3 I see two options:

1. Change babashka to infer `:paths ["src"]` when `bb` is invoked with `--config`.
2. Require `bbin` script authors to set `:paths` explicitly if they want to use `bbin` to ship their scripts.

(See heading "Question for @borkdude: should classpath be inferred for `bb --config some-deps-edn-or-bb-edn-file`?" in linked document.)

**Comment by borkdude** [comment]
With bb `:paths` is always explicit, as to not mix JVM clojure sources with bb sources, so explicit seems best to me.

**Comment by rads** [comment]
@teodorlu: It's been a while since I looked at this code in detail so I appreciate the overview. You've already started moving towards the right direction in your PR, but I'll summarize my suggested plan here for clarity:

- The following changes are only relevant to the `GitDir` and `LocalDir` script types:
  - If `deps.edn` exists, continue using `--deps-root` and `--config`.
    - This ensures that all existing projects that are already working maintain exactly the same behavior as before, even if reinstalled with the new verison of `bbin`.
  - If `deps.edn` does not exist, look for `bb.edn` and use only `--config`.
    - This case was broken anyways which means it's safe to add new behavior here.
    - The `:paths` in the `bb.edn` **must** be set explicitly, otherwise throw an exception.
  - If neither `deps.edn` nor `bb.edn` exist, throw an exception.
- It would be ideal for `bbin` to provide friendly error messages when it detects these cases rather than letting the `bb` command fail.

If the above points look good to you and @borkdude, then we can move forward with finalizing the implementation.

**Comment by borkdude** [comment]
Looks good, except maybe for this one?

> The :paths in the bb.edn must be set explicitly, otherwise throw an exception.

Why throw an exception in this case, one could have a valid `bb.edn` without `:paths`, e.g. with a local/root dep or whatever

**Comment by rads** [comment]
Good point. I'll fix that to say that we don't want to add inference rather than adding a strict check.

**Comment by borkdude** [comment]
@teodorlu 

> If the above points look good to you and @borkdude, then we can move forward with finalizing the implementation.

I think it all looks good, except the throwing.

**Comment by teodorlu** [comment]
I think that it might be a good idea to move away from string-based selmer templates for the shims to code as data.

1. With code as data, we can get structural editing and syntax highlighting from the editor
2. With code as data, we can also avoid duplicating code, not have to solve the same problem in two places.

To get pretty-printed code, we could use something like `clojure.pprint/code-dispatch` ([example](https://clojuredocs.org/clojure.pprint/with-pprint-dispatch#example-608ec17ce4b0b1e3652d74ef)).

@rads / @borkdude thoughts on this?

(Just saying this now, because I felt it made the last "chunk" of work harder than necessary)

**Comment by borkdude** [comment]
I'm fine with this, but it seems like a different issue to me, let's not try to do too many changes at once, unless it makes your life easier in the PR?

**Comment by borkdude** [comment]
@teodorlu What's necessary to wrap this issue up? I think it was mentioned in the bbin channel on Clojurians slack once again

**Comment by teodorlu** [comment]
Hey!

It's been a while since I last looked at this. I lost a bit of steam working on the Clojure code we generate by interpolating values into strings. I also remember feeling slightly uneasy changing the generated Clojure code fearing Windows regressions. Maybe first refactoring to generate Clojure code as Clojure data is easier.

If you want to take a stab (possibly you'll just steam through an hour as you sometimes do), feel free!

Otherwise I'll give it another shot in a day or two.

**Comment by rads** [comment]
@borkdude: As part of fixing this issue, I'm planning to change the underlying command used in the generated scripts:
```
# Before
bb --config path/to/project/bb.edn --deps-root path/to/project -- ...

# After (if bb.edn exists)
bb --config path/to/project/bb.edn -- ...

# After (if bb.edn doesn't exist)
bb -- ...
```
This means no `deps.edn` is necessary and the behavior will be the same as if you were running `bb` from the project directory. However, it is not a completely backwards-compatible change when installing new scripts.

For example, with the current verison of bbin, you can install the parmezan repo using the following command:
```
bbin install https://github.com/borkdude/parmezan.git
```
Since the installed bbin script is using `--deps-root`, the `src` path gets added to the classpath implicitly. However, you can't actually run `bb -m borkdude.parmezan.cli` from within the repo itself because `src` isn't on the classpath in `bb.edn` or `deps.edn`.

I think bbin scripts should use the `bb.edn` as the source of truth going forward to keep things as consistent as possible between running Babashka code with and without bbin. You can still use a `deps.edn` using the existing idiom of adding `{:local/root "."}` as a dependency in the project `bb.edn`. Even though this changes how the generated scripts work, it only applies to newly installed scripts, so it doesn't seem like a big deal to change this as long as we document this in the next version bump.

Let me know if you have any thoughts about this.

**Comment by rads** [comment]
Although now that I'm thinking about it some more, changing the install behavior means that some existing projects that could be installed in bbin 0.2.x would no longer be able to be installed without updates in bbin 0.3.x. So we probably need some way to handle existing repos that relied on this implicit behavior (assuming we want to use `bb.edn` as the source of truth for new installs).

**Comment by borkdude** [comment]
`bb.edn` is often used for project tasks so it's not always the source of truth for an installable script either?

**Comment by rads** [comment]
It's also possible for the Babashka and Clojure code to be completely separate, even if the `bb.edn` and `deps.edn` are in the same project root. Should the `deps.edn` take precedence over the `bb.edn` when it exists? Given that bbin is a tool to install _Babashka_ scripts, that seems confusing, especially given that there are existing patterns for using a `deps.edn` file from Babashka.

If someone is using `bb.edn` in a Clojure project for tasks and they want to install it via Babashka, it seems like they should set up their project to be usable from Babashka first without relying on bbin to implicitly fix it for them. I can see an argument for being able to pass custom `bb` flags (such as `--deps-root`) as part of installing a script, but that seems better as an explicit config rather than implicit behavior.

**Comment by borkdude** [comment]
Thinking more about it: you're probably right. Another option could be to put some explicit configuration in a `bbin.edn` file and use `--config bbin.edn` instead. But `bb.edn` probably is a sane default. I should add `src` to my `bb.edn` to allow bbin to run it, no problem with that.
